sw/device/tests/sram_ctrl_main_scrambled_access_test.c

```diff
@@ -0,0 +1,177 @@
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+            0x5A5A5A5A,
+        },
+};
+
+/**
+ * Performs scrambling, saves the test relevant data and resets the system.
+ *
+ * MAIN SRAM will be overwritten with the pseudo-random data (including .bss
+ * and .data segments), which means that the code must be written in assembly.
+ *
+ * This function saves contents of the `sram_ctrl_main_scramble_buffer` in the
+ * RETENTION SRAM, which maintains the power across the system reboot.
+ */
+static noreturn void main_sram_scramble(void) {
+  asm volatile(
+      // Request the new scrambling key for MAIN SRAM.
+      "li %[kGeneralRW], 0x1;                                              \n"
+      "sw %[kGeneralRW], %[kSramCtrlOffset](%[kSramCtrlRegsBase]);         \n"
+
+      // Busy loop - waiting for scrambling to finish.
+      ".L_SCRAMBLING_BUSY_LOOP:;                                           \n"
+      "  lw %[kGeneralRW], %[kSramCtrlStatusOffset](%[kSramCtrlRegsBase]); \n"
+      "  andi %[kGeneralRW], %[kGeneralRW], %[kSramCtrlKeyScrDone];        \n"
+      "  beqz %[kGeneralRW], .L_SCRAMBLING_BUSY_LOOP;                      \n"
+
+      // Copy the buffer contents to the retention SRAM.
+      ".L_BUFFER_COPY_LOOP:;                                               \n"
+      "  lw %[kGeneralRW], 0(%[kCopyFromAddr]);                            \n"
+      "  sw %[kGeneralRW], 0(%[kCopyToAddr]);                              \n"
+      "  addi %[kCopyFromAddr], %[kCopyFromAddr], 4;                       \n"
+      "  addi %[kCopyToAddr], %[kCopyToAddr], 4;                           \n"
+      "  blt %[kCopyToAddr], %[kCopyToEndAddr], .L_BUFFER_COPY_LOOP;       \n"
+
+      // Trigger the software system reset via the Reset Manager.
+      "li %[kGeneralRW], %[kMultiBitTrue];                                 \n"
+      "sw %[kGeneralRW], %[kRstMgrResetReq](%[kRstMgrRegsBase]);           \n"
+
+      // Satisfy the `noreturn` promise to the compiler.
+      ".L_INFINITE_LOOP:;                                                  \n"
+      "  wfi;                                                              \n"
+      "  j .L_INFINITE_LOOP;"
+      :
+      : [kGeneralRW] "r"(0), [kMultiBitTrue] "I"(kMultiBitBool4True),
+
+        [kSramCtrlRegsBase] "r"(TOP_EARLGREY_SRAM_CTRL_MAIN_REGS_BASE_ADDR),
+        [kSramCtrlOffset] "I"(SRAM_CTRL_CTRL_REG_OFFSET),
+        [kSramCtrlStatusOffset] "I"(SRAM_CTRL_STATUS_REG_OFFSET),
+        [kSramCtrlKeyScrDone] "I"(0x1 << SRAM_CTRL_STATUS_SCR_KEY_VALID_BIT),
+
+        [kRstMgrRegsBase] "r"(TOP_EARLGREY_RSTMGR_AON_BASE_ADDR),
+        [kRstMgrResetReq] "I"(RSTMGR_RESET_REQ_REG_OFFSET),
+
+        [kCopyFromAddr] "r"(sram_ctrl_main_scramble_buffer),
+        [kCopyToAddr] "r"(TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR),
+        [kCopyToEndAddr] "r"(TOP_EARLGREY_SRAM_CTRL_RET_AON_RAM_BASE_ADDR +
+                             SRAM_CTRL_TESTUTILS_DATA_NUM_WORDS));
+
+  __builtin_unreachable();
```

Comment:
Is there an OT_UNREACHABLE macro? If not can you add one?

Reply:
Done.